plugins/TrollManagement/class.trollmanagement.plugin.php

```diff
@@ -469,4 +505,201 @@ private function checkTroll(int $userID, array &$args) {
+            $maxSiblingAccounts = c('TrollManagement.PerFingerPrint.MaxUserAccounts');
+            if ($userFingerprint !== false) {
+                if ($this->checkMaxSharedFingerprintsExceeded($userFingerprint, $maxSiblingAccounts)) {
+                    Gdn::userModel()->addRoles($userID, [RoleModel::APPLICANT_ID], true);
+                }
+            }
+        }
+    }
+
+    /**
+     * Adds "Fingerprint" to the allowed banTypes of the "Ban Rules" dashboard interface.
+     *
+     * @param array $banTypes
+     * @return array
+     */
+    public function settingsController_listBanTypes(array $banTypes): array {
+        $banTypes['Fingerprint'] = t('Fingerprint');
+        return $banTypes;
+    }
+
+    /**
+     * Either add "Fingerprint" row header or an fingerprint value to the dashboard's list of users.
+     *
+     * @param Controller $sender
+     * @param array $args
+     */
+    public function base_userCell_handler($sender, $args) {
+        // If we have user data, we create a cell containing the fingerprint.
+        if (isset($args['User'])) {
+            echo '<td>' . val('Fingerprint', $args['User']) . '</td>';
+        } else {
+            // Otherwise, we output a column header.
+            $get = Gdn::request()->get();
+            $get['order'] = 'Fingerprint';
+            $orderUrl = '/dashboard/user?'.http_build_query($get);
+
+            echo '<th class="column-md">' . anchor(t('Fingerprint'), $orderUrl) . '</th>';
+        }
+    }
+
+    /**
+     * Add "Fingerprint" to the possible ban query.
+     *
+     * @param array $result
+     * @param array $ban
+     * @return array
+     */
+    public function banModel_banWhere_handler(array $result, array $ban): array {
+        switch (strtolower($ban['BanType'])) {
+            case 'fingerprint':
+                $result['u.Fingerprint'] = $ban['BanValue'];
+                break;
+        }
+        return $result;
+    }
+
+    /**
+     * Add "Fingerprint" to the dashboard's users list search query.
+     *
+     * @param array $whereCriterias
+     * @param string $keywords
+     * @return array
+     */
+    public function userModel_searchKeyWords_handler(array $whereCriterias, string $keywords): array {
+        $whereCriterias['where']['u.Fingerprint'] = $keywords;
+
+        return $whereCriterias;
+    }
+
+    /**
+     * Add ordering users by Fingerprint in the dashboard's users list.
+     *
+     * @param array $allowedSorting
+     * @return array
+     */
+    public function userController_usersListAllowedSorting(array $allowedSorting): array {
+        $allowedSorting['Fingerprint'] = 'desc';
+        return $allowedSorting;
+    }
+}
+
+if (!function_exists('validatePositiveNumber')) {
+    /**
+     * Validate if number is positive
+     *
+     * @param int|string $number
+     * @return bool
+     */
+    function validatePositiveNumber($number): bool {
+        return (is_numeric($number) && (int)$number > 0);
+    }
```

Comment:
Was this copied & pasted from the QnA addon? If it's needed in multiple places, it's something we should be adding to core.

Reply:
Agreed. Moved it to functions.validation.php